plugins/jvm-abi-gen/src/org/jetbrains/kotlin/jvm/abi/JvmAbiMetadataProcessor.kt

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.jvm.abi
+
+import kotlinx.metadata.*
+import kotlinx.metadata.jvm.JvmClassExtensionVisitor
+import kotlinx.metadata.jvm.JvmPackageExtensionVisitor
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import org.jetbrains.kotlin.load.java.JvmAnnotationNames.*
+import org.jetbrains.org.objectweb.asm.AnnotationVisitor
+import org.jetbrains.org.objectweb.asm.Opcodes
+
+/**
+ * Wrap the visitor for a Kotlin Metadata annotation to strip out private and local
+ * functions, properties, and type aliases as well as local delegated properties.
+ */
+fun abiMetadataProcessor(annotationVisitor: AnnotationVisitor): AnnotationVisitor =
+    kotlinClassHeaderVisitor { header ->
+        // kotlinx-metadata only supports writing Kotlin metadata of version >= 1.4, so we need to
+        // update the metadata version if we encounter older metadata annotations.
+        val metadataVersion =
+            if (header.metadataVersion.let { it.size >= 2 && it[0] >= 1 && it[1] >= 4 }) header.metadataVersion else intArrayOf(1, 4)
+        val newHeader = when (val metadata = KotlinClassMetadata.read(header)) {
+            is KotlinClassMetadata.Class -> {
+                val writer = KotlinClassMetadata.Class.Writer()
+                metadata.accept(AbiKmClassVisitor(writer))
```

Comment:
What about private classes (top level,  nested)?

Reply:
How they skipped, processed?

Reply:
Right now they are kept in the metadata. The reasoning is that since we don't strip private classes from the ABI (as they are still visible from Java) it's probably for the best to keep them in the metadata as well. Otherwise they wouldn't be recognized as nested classes at all, which could lead to further problems.

Reply:
Is it any test for that?

Reply:
maybe with Java client

Reply:
Good point, I've added tests for private (annotation) classes in ba5417d.